File: velocity/src/main/java/ru/privatenull/pnauth/velocity/VelocityAuthActions.java
package ru.privatenull.pnauth.velocity;

import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.server.RegisteredServer;
import ru.privatenull.pnauth.command.AuthPlatformBridge;
import ru.privatenull.pnauth.config.ProxySettings;
import ru.privatenull.pnauth.message.AuthMessages;
import ru.privatenull.pnauth.message.MessageFormat;

import java.util.UUID;

final class VelocityAuthActions implements AuthPlatformBridge {
    private final ProxyServer proxy;
    private final ProxySettings settings;
    private final AuthMessages messages;
    private final MessageFormat messageFormat;

    VelocityAuthActions(ProxyServer proxy, ProxySettings settings, AuthMessages messages, MessageFormat messageFormat) {
        this.proxy = proxy;
        this.settings = settings;
        this.messages = messages;
        this.messageFormat = messageFormat;
    }

    @Override
    public void authenticated(UUID uniqueId) {
        Player player = proxy.getPlayer(uniqueId).orElse(null);
        authenticated(player);
    }

    @Override
    public void authenticated(String username) {
        authenticated(proxy.getPlayer(username).orElse(null));
    }

    private void authenticated(Player player) {
        if (player == null) return;
        if (!settings.hasBackendServer()) return;
        // Initial routing is applied by PlayerChooseInitialServerEvent from the shared lifecycle decision.
        // Authentication events only move players that are already connected to an auth server.
        if (player.getCurrentServer().isEmpty()) return;
        String serverName = player.getVirtualHost()
                .map(host -> settings.forcedHosts().getOrDefault(host.getHostString().toLowerCase(), settings.backendServer()))
                .orElse(settings.backendServer());
        RegisteredServer target = proxy.getServer(serverName).orElse(null);
        if (target == null) {
            player.disconnect(VelocityMessages.component(messages.text("access.backend_missing"), messageFormat));
            return;
        }
        if (player.getCurrentServer().map(server -> !server.getServer().equals(target)).orElse(true)) {
            player.createConnectionRequest(target).connect();
        }
    }

    @Override
    public void loggedOut(UUID uniqueId) {
        proxy.getPlayer(uniqueId).ifPresent(player -> player.disconnect(
                VelocityMessages.component(messages.text("logout.disconnect"), messageFormat)));
    }

    @Override
    public void accountDeleted(UUID uniqueId) {
        proxy.getPlayer(uniqueId).ifPresent(player -> player.disconnect(
                VelocityMessages.component(messages.text("unregister.disconnect"), messageFormat)));
    }

    @Override
    public void accountDeleted(String username) {
        proxy.getPlayer(username).ifPresent(player -> player.disconnect(
                VelocityMessages.component(messages.text("unregister.disconnect"), messageFormat)));
    }

    @Override
    public void broadcast(String message) {
        proxy.getAllPlayers().forEach(player -> player.sendMessage(
                VelocityMessages.component(messages.text("broadcast.message", java.util.Map.of("message", message)), messageFormat)));
    }
}

File: velocity/src/main/kotlin/ru/privatenull/pnauth/velocity/VelocityAuthActions.kt
package ru.privatenull.pnauth.velocity

import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import ru.privatenull.pnauth.config.ProxySettings
import ru.privatenull.pnauth.message.AuthMessages
import ru.privatenull.pnauth.message.MessageFormat
import ru.privatenull.pnauth.platform.Proxy
import ru.privatenull.pnauth.routing.ServerBalancerFactory
import java.util.UUID

internal class VelocityAuthActions(
    private val proxy: ProxyServer,
    private val settings: ProxySettings,
    private val messages: AuthMessages,
    private val messageFormat: MessageFormat,
    private val proxyAdapter: Proxy? = null
) : ru.privatenull.pnauth.platform.adapter.PlatformAuthBridgeAdapter {

    override fun authenticated(uniqueId: UUID) {
        authenticated(uniqueId, false)
    }

    override fun authenticated(uniqueId: UUID, isRegistration: Boolean) {
        val player = proxy.getPlayer(uniqueId).orElse(null)
        authenticated(player, isRegistration)
    }

    override fun authenticated(username: String) {
        authenticated(proxy.getPlayer(username).orElse(null), false)
    }

    private fun authenticated(player: Player?, isRegistration: Boolean = false) {
        if (player == null) return

        // Display success Title & Subtitle with gradient
        val titleKey = if (isRegistration) "title.register.success" else "title.login.success"
        val subtitleKey = if (isRegistration) "subtitle.register.success" else "subtitle.login.success"
        val titleComp = VelocityMessages.component(messages.text(titleKey), messageFormat)
        val subtitleComp = VelocityMessages.component(messages.text(subtitleKey), messageFormat)
        val titleObj = net.kyori.adventure.title.Title.title(
            titleComp,
            subtitleComp,
            net.kyori.adventure.title.Title.Times.times(
                java.time.Duration.ofMillis(500),
                java.time.Duration.ofMillis(2000),
                java.time.Duration.ofMillis(500)
            )
        )
        player.showTitle(titleObj)

        if (!settings.hasBackendServer()) return
        if (player.currentServer.isEmpty) return

        var targets = settings.getEffectiveBackendServers()
        val virtualHost = player.virtualHost.orElse(null)
        if (virtualHost != null) {
            val forced = settings.forcedHosts[virtualHost.hostString.lowercase()]
            if (forced != null) targets = listOf(forced)
        }
        val balancer = ServerBalancerFactory.create(settings.balancerMode, settings.maxPlayersPerServer, settings.serverLimits)
        val selected = balancer.selectServer(targets, proxyAdapter).orElse(settings.backendServer)
        val target = proxy.getServer(selected).orElse(null)

        if (target == null) {
            player.disconnect(VelocityMessages.component(messages.text("access.backend_missing"), messageFormat))
            return
        }
        if (player.currentServer.map { server -> server.server != target }.orElse(true)) {
            player.createConnectionRequest(target).connect()
        }
    }

    override fun loggedOut(uniqueId: UUID) {
        proxy.getPlayer(uniqueId).ifPresent { player ->
            player.disconnect(VelocityMessages.component(messages.text("logout.disconnect"), messageFormat))
        }
    }

    override fun accountDeleted(uniqueId: UUID) {
        proxy.getPlayer(uniqueId).ifPresent { player ->
            player.disconnect(VelocityMessages.component(messages.text("unregister.disconnect"), messageFormat))
        }
    }

    override fun accountDeleted(username: String) {
        proxy.getPlayer(username).ifPresent { player ->
            player.disconnect(VelocityMessages.component(messages.text("unregister.disconnect"), messageFormat))
        }
    }

    override fun broadcast(message: String) {
        proxy.allPlayers.forEach { player ->
            player.sendMessage(
                VelocityMessages.component(
                    messages.text("broadcast.message", mapOf("message" to message)),
                    messageFormat
                )
            )
        }
    }
}
